crates/sui-data-ingestion/src/workers/archival.rs

```diff
@@ -138,67 +112,69 @@ impl ArchivalWorker {
-                .iter()
-                .map(|t| ExecutionData::new(t.transaction.clone(), t.effects.clone())),
-        );
-        let contents_blob = Blob::encode(&full_checkpoint_contents, BlobEncoding::Bcs)?;
-        let blob_size = contents_blob.size();
-        let summary_blob = Blob::encode(&checkpoint.checkpoint_summary, BlobEncoding::Bcs)?;
-
-        if !state.buffer.is_empty()
-            && (((state.buffer.len() + blob_size) > self.commit_file_size)
-                || state.epoch != epoch
-                || checkpoint.checkpoint_summary.timestamp_ms
-                    > (self.commit_duration_ms + state.last_commit_ms))
-        {
-            self.upload(&state).await?;
-            state.epoch = epoch;
-            state.checkpoint_range = sequence_number..sequence_number;
-            state.buffer = vec![];
-            state.summary_buffer = vec![];
-            state.last_commit_ms = checkpoint.checkpoint_summary.timestamp_ms;
-            state.should_update_progress = true;
+        let mut summary_buffer = vec![];
+        let mut buffer = vec![];
+        let first_checkpoint = &batch[0];
+        let epoch = first_checkpoint.checkpoint_summary.epoch;
+        let start_checkpoint = first_checkpoint.checkpoint_summary.sequence_number;
+        let mut last_checkpoint = start_checkpoint;
+        for checkpoint in batch {
+            let full_checkpoint_contents = FullCheckpointContents::from_contents_and_execution_data(
+                checkpoint.checkpoint_contents.clone(),
+                checkpoint
+                    .transactions
+                    .iter()
+                    .map(|t| ExecutionData::new(t.transaction.clone(), t.effects.clone())),
+            );
+            let contents_blob = Blob::encode(&full_checkpoint_contents, BlobEncoding::Bcs)?;
+            let summary_blob = Blob::encode(&checkpoint.checkpoint_summary, BlobEncoding::Bcs)?;
+            contents_blob.write(&mut buffer)?;
+            summary_blob.write(&mut summary_buffer)?;
+            last_checkpoint += 1;
         }
-        contents_blob.write(&mut state.buffer)?;
-        summary_blob.write(&mut state.summary_buffer)?;
-        state.checkpoint_range.end += 1;
+        self.upload(
+            epoch,
+            start_checkpoint,
+            last_checkpoint,
+            summary_buffer,
+            buffer,
+        )
+        .await?;
         Ok(())
     }
 
-    async fn save_progress(
+    fn should_close_batch(
         &self,
-        sequence_number: CheckpointSequenceNumber,
-    ) -> Option<CheckpointSequenceNumber> {
-        let mut state = self.state.lock().await;
-        let should_update_progress = state.should_update_progress;
-        state.should_update_progress = false;
-        if should_update_progress && sequence_number > 0 {
-            Some(sequence_number - 1)
-        } else {
-            None
+        batch: &[CheckpointData],
+        next_item: Option<&CheckpointData>,
```

Comment:
I am actually wondering if the interface could be simplified to not having to pass `next_item`.
In this case, you would just need to look into the last `CheckpointData` and see if it contains the end_of_epoch data. I don't know if you care about the slight timestamp difference between two checkpoints for the timestamp check, but if that's tolerable, I think most likely all other pipelines don't need to look at the next item to make a commit decision.

Reply:
Yes, I guess that would work for archival. Timestamps being slightly off is totally fine. 
Not sure if that would be enough for the general case

Reply:
Could you think of a case where we need the next item though?

Reply:
not really. So we can go with a simplified interface for now and iterate on top if necessary

Reply:
but then how do we differentiate inside of the `should_close_batch` function between an [internal](https://github.com/MystenLabs/sui/pull/19656/files#diff-af5aaddb6e99077a7d722eea57fa30a40500f06020a4b872d6ef877a9bd481aaR32) iteration of the loop and a [final](https://github.com/MystenLabs/sui/pull/19656/files#diff-af5aaddb6e99077a7d722eea57fa30a40500f06020a4b872d6ef877a9bd481aaR44) one

Reply:
Do we need to though? I think the same condition can apply.

Reply:
for archival, it’s the same because it has a strict rule for when to seal a batch. 
For the PG primary workflow, `should_close_batch` should return false during internal iteration to avoid sealing the batch prematurely, and `true` in the final iteration because we want to commit whatever we've accumulated so far

Reply:
Ah you are right!